Dockerfile

```diff
@@ -30,3 +30,26 @@ EXPOSE 5000
 ENTRYPOINT [ "./run.sh", "dev"]
 RUN poetry install --no-root
 
+FROM base as test
+
+# Install Chrome
+RUN curl -sSL https://dl.google.com/linux/direct/google-chrome-stable_current_amd64.deb -o chrome.deb &&\
+    apt-get update &&\
+    apt-get install ./chrome.deb -y --fix-missing &&\
+    rm ./chrome.deb
+# Install Chromium WebDriver
+RUN LATEST=`curl -sSL https://chromedriver.storage.googleapis.com/LATEST_RELEASE` &&\
+echo "Installing chromium webdriver version ${LATEST}" &&\
+curl -sSL https://chromedriver.storage.googleapis.com/${LATEST}/chromedriver_linux64.zip -o chromedriver_linux64.zip &&\
+apt-get install unzip -y &&\
+unzip ./chromedriver_linux64.zip
+
+RUN poetry install --no-root
+COPY tests.sh ./tests.sh
+RUN chmod +x ./tests.sh
+COPY tests ./tests
+COPY integration_tests ./integration_tests
+COPY e2e_tests ./e2e_tests
+COPY todo_app ./todo_app

```

Comment:
This could be a lot simpler if you just `COPY . .`
What are you trying so hard to avoid copying? Could you add those things to dockerignore instead?